---
title: "Retrofit Vignette"
author: "Roopali Singh"
date: "`r Sys.Date()`"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Retrofit Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Introduction

RETROFIT is a statistical method for reference-free deconvolution of spatial transcriptomics data to estimate cell type mixtures. In this Vignette, we will estimate cell type composition of a sample synthetic dataset. We will annotate cell types using an annotated single cell reference.

## Package Installation

Install and load the package using the following steps:
<!-- adam: this is not usable before submitting to and accepted by Bioconductor -->
```{r, eval=FALSE}
if(!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("retrofit")
```

```{r, load_library}
library(retrofit)
```

## Spatial Transcriptomics Data

First load the ST data, using the following command:

```{r, data}
X         =read.csv("data/sample_x.csv", row.names = 1, check.names = FALSE)
```
The ST data matrix will consist of G = 500 genes and S = 1000 spots i.e., a matrix of order G x S.

## Reference-free Deconvolution

Initialize the following parameters for deconvolution:
Iter: Number of iterations (default = 4000)
L: Number of components required
Lamda (default value is 0.01)

After initialization, run retrofit on the data (X) as follows:

```{r, decompose}
Iter = 4000
L = 20
result = RetrofitDecompose(X, L=L, iterations=Iter, seed=12)
H = result["h"]
W = result["w"]
Theta =result["th"]
```

## Cell-type Annotation via annotated single cell reference

After deconvolution of ST data, we have our estimates of W (a matrix of order G x L), H (a matrix of order L x S) and Theta (a vector of L components). 
Next, we need to annotate the components, to get the proportion of, say K, cell types. We can do this in two ways: (a) using an annotated single cell reference or (b) using the known marker genes. Here, we will annotate using single cell reference.

Load the single cell reference data:

```{r, data}
sc_ref   =read.csv("data/sc_ref.csv", row.names = 1, check.names = FALSE)
```
This file contains average gene expression values for G genes in K cell types i.e., a matrix of order G x K. Run the following command to get K cell-type mixtures from the ST data X:

```{r, annotate}
K = ncol(W) # number of cell types
result = RetrofitMapByCorrelation(sc_ref, K, W, H)
H_annot = result["h"]
W_annot = result["w"]
```
[Visualization here] 

Above code assigns components to the cell type it has maximum correlation with, as shown in the plot above.


## Deconvolution with annotation (Optional)

We can also deconvolve the ST data matrix along with cell-type annotation all in one step, as follows:

```{r, run_retrofit}
Iter = 4000
K = 10 # number of cell types in reference
L = 20 # number of components
result = RetrofitMain(X, 
                      ref_cor = sc_ref,
                      iterations=Iter, 
                      L=L, 
                      K=K,
                      seed=12)
```

## Results and visualization

Since this is synthetic data, true cell type proportions are known. We can visualize the correlation and RMSE between true and estimated cell-type proportions as follows:

<!-- Adam: need the code to create RMSE and correlation plot for proportions i.e., cor(H, H_hat) and RMSE(H, H_hat) -->

```{r}
# correlation between true and estimated cell-type proportions
library(ggplot2)
reference=as.matrix(read.csv("data/sc_ref.csv", row.names=1, check.names=FALSE))
result=read.csv("data/sample_results/sample_x__map_cor_w.csv", row.names=1, check.names=FALSE)

correlations=sort(diag(cor(reference, result)), decreasing=T, na.last=T)
df=data.frame(x=1:length(correlations), 
              y=correlations, 
              label_x1=1, 
              label_x2=2, 
              label_y=seq(from=0.5, by=-0.05, length.out=10),
              label_cell=names(correlations),
              label_corr=format(round(correlations, digits=4)))

gg <- ggplot(df,aes(x=x, y=y, group=1)) + 
  geom_line(aes(x=x, y=y)) + geom_point(aes(x=x, y=y)) + theme_bw() + ylim(0, 1.05) +
  geom_text(data=df, aes(x=label_x1, y=label_y, label=label_corr), size=4, hjust=0) +
  geom_text(data=df, aes(x=label_x2, y=label_y, label=label_cell), size=4, hjust=0)
plot(gg)
```

```{r}
# RMSE between true and estimated cell-type proportions
library(ggplot2)
H_True=read.csv("data/sc_ref_h.csv", row.names=1, check.names=FALSE)
H=read.csv("data/sample_results/sample_x__map_cor_h.csv", row.names=1, check.names=FALSE)

rmse=NULL
for(i in 1:ncol(H_True)){
  H_True_i=H_True[,i]/sum(H_True[,i])
  H_i=H[,i]/sum(H[,i])
  rmse=c(rmse,sqrt(mean((H_True_i-H_i)^2)))
}
df=data.frame(x=1, y=rmse)

gg <- ggplot(df,aes(x=x,y=y, fill=x)) + 
  geom_violin(lwd=0) + theme_bw() + xlim(0,5) + ylim(0, 0.75) +
  theme(legend.position = "None",
        axis.title=element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_text(size=10))
plot(gg)
```

In case of real data, where ground truth is not available, we can do certain diagnostics on the quality of results. For example, compute the correlation between marker expression (if known) and cell type proportion in ST data and visualize the agreement of the spatial pattern between markers and estimated proportion (read the paper for more information).

# Session information

```{r}
sessionInfo()
```



